pkg/util/liveconfig/hive.go

```diff
@@ -41,37 +81,32 @@ func parseKubeconfig(credentials []mgmtcontainerservice.CredentialResult) (*rest
 	return restConfig, nil
 }
 
-func (p *prod) HiveRestConfig(ctx context.Context, index int) (*rest.Config, error) {
-	// NOTE: This RWMutex locks on a fetch for any index for simplicity, rather
-	// than a more granular per-index lock. As of the time of writing, multiple
-	// Hive shards are planned but unimplemented elsewhere.
+func (p *prod) HiveRestConfig(ctx context.Context, index int, credentialType AksCredentialType) (*rest.Config, error) {
+	// Hive shards are planned but not implemented yet
 	p.hiveCredentialsMutex.RLock()
-	cached, ext := p.cachedCredentials[index]
-	p.hiveCredentialsMutex.RUnlock()
-	if ext {
-		return rest.CopyConfig(cached), nil
+	credentialCache, exists := p.cachedCredentials[credentialType]
+	if exists {
+		cached, exists := credentialCache[index]
+		p.hiveCredentialsMutex.RUnlock()
+		if exists {
+			return rest.CopyConfig(cached), nil
+		}
+	} else {
+		p.hiveCredentialsMutex.RUnlock()
 	}
 
-	// Lock the RWMutex as we're starting to fetch so that new readers will wait
-	// for the existing Azure API call to be done.
-	p.hiveCredentialsMutex.Lock()
-	defer p.hiveCredentialsMutex.Unlock()
-
-	rpResourceGroup := fmt.Sprintf("rp-%s", p.location)
-	rpResourceName := fmt.Sprintf("aro-aks-cluster-%03d", index)
-
-	res, err := p.managedClustersClient.ListClusterUserCredentials(ctx, rpResourceGroup, rpResourceName, "")
+	kubeConfig, err := getAksKubeconfig(ctx, p.managedClustersClient, p.location, index, credentialType)
 	if err != nil {
 		return nil, err
 	}
 
-	parsed, err := parseKubeconfig(*res.Kubeconfigs)
-	if err != nil {
-		return nil, err
+	p.hiveCredentialsMutex.Lock()
```

Comment:
Consideration: Moving the lock here reduces the time we block the individual goroutines but will likely see multiple calls to the Azure API. We add random delays up to a minute at monitor goroutines creation time, so the number of calls are a function of the number of clusters a monitor instance is handling (currently ~100 max) and the runtime of the Azure API calls, which should be below 500 ms => if the randomisation is good enough we should be fine and not trigger denial of service attacks to the Azure API.

Reply:
Consideration 2: If all calls fail (current situation) and the cache remains empty every monitor goroutine will trigger an API Call every minute, so we'll create 1 to 2 call/sec, which still should be fine.

Reply:
I went back to the original, to avoid any authentication "stampedes". Since we already have the cluster admin role assigned to the RP MSI for all of the AKS clusters in production, we aren't going to see the level of credential requests we have been and we should be fine from that aspect.

The only consideration is that the context is [only set to 50 seconds](https://github.com/Azure/ARO-RP/blob/aacda3a665eca2bf7faa9f9b768c63d20951f95a/pkg/monitor/worker.go#L192) so in cases where Azure does respond slowly we may drop some metrics for a collection cycle, but it should only be once.

Reply:
I have refactored the method of fetching the kubeConfigs, taking to a global level, because we only need to do it once for all monitors and yet it was in the critical monitoring path of all of them.